uefi/section.go

```diff
@@ -135,13 +177,105 @@ func (s *Section) ApplyChildren(v Visitor) error {
 	return nil
 }
 
-// Assemble assembles the section from the binary
-func (s *Section) Assemble() ([]byte, error) {
+// GenSecHeader generates a full binary header for the section data.
+// It assumes that the passed in section struct already contains section data in the buffer,
+// the section type in the Type field, and the type specific header in the TypeSpecific field.
+// It modifies the calling Section.
+func (s *Section) GenSecHeader() error {
 	var err error
-	s.buf, err = ioutil.ReadFile(s.ExtractPath)
+	// Calculate size
+	headerLen := uint32(SectionMinLength)
+	if s.TypeSpecific != nil && s.TypeSpecific.Header != nil {
+		headerLen += s.TypeSpecific.Header.GetBinHeaderLen()
+	}
+	log.Print(headerLen)
+	s.Header.ExtendedSize = uint32(len(s.buf)) + headerLen // TS header lengths are part of headerLen at this point
+	if s.Header.ExtendedSize > 0xFFFFFF {
+		headerLen += 4 // Add space for the extended header.
+		s.Header.ExtendedSize += 4
+	}
+
+	// Set the correct data offset for GUID Defined headers.
+	// This is terrible
+	if s.Header.Type == SectionTypeGUIDDefined {
+		gd := s.TypeSpecific.Header.(*SectionGUIDDefined)
+		gd.DataOffset = uint16(headerLen)
+		// append type specific header in front of data
+		tsh := new(bytes.Buffer)
+		if err = binary.Write(tsh, binary.LittleEndian, &gd.SectionGUIDDefinedHeader); err != nil {
+			return err
+		}
+		s.buf = append(tsh.Bytes(), s.buf...)
+	}
+
+	// Append common header
+	s.Header.Size = Write3Size(uint64(s.Header.ExtendedSize))
+	h := new(bytes.Buffer)
+	if s.Header.ExtendedSize > 0xFFFFFF {
```

Comment:
Shouldn't this be `>=`? If the section size is exactly 0xFFFFFF, it would have to use the extended header.

Reply:
Done